### PR TITLE
Introduce the memquota2 adapter.

### PR DESCRIPTION
**Pull request by geeknoid**
This is a cut-and-paste of the original memory-based quota adapter. The core quota logic is unchanged, it's just the glue with the rest of the system that is new.

100% code coverage.

**Release note**:

```release-note
```

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/1156)

### CODE REVIEW

File: adapter/memquota2/config/config.proto [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import "google/protobuf/duration.proto";
+import "gogoproto/gogo.proto";
+
+option go_package="config";
+option (gogoproto.goproto_getters_all) = false;
+option (gogoproto.equal_all) = false;
+option (gogoproto.gostring_all) = false;
+
+message Params {
+	message Quota {
```
Comment (ZackButcher) [inline adapter/memquota2/config/config.proto]: I've been calling these sub-messages "FooInfo". I think it'd be better if these were named consistently across all of our adapters for this naming. Do we want to call these "Foo" or "FooInfo"? My concern is confusion between instances and these objects if we call them "Foo" (e.g. "Quota" is the name we use for the Istio-thing, and also the name you use in this adapter, but they're not the same thing).

File: adapter/memquota2/memquota.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+func (h *handler) Close() error {
+	h.common.ticker.Stop()
+	return nil
+}
+
+////////////////// Bootstrap //////////////////////////
+
+// GetBuilderInfo returns the BuilderInfo associated with this adapter implementation.
+func GetBuilderInfo() adapter.BuilderInfo {
```
Comment (ZackButcher) [inline adapter/memquota2/memquota.go]: Nit: move this to the top of the file. It's effectively the summary of what's going on in this adapter, I think the whole file reads better with this up top.

File: adapter/memquota2/memquota.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+func (b *builder) ConfigureQuotaHandler(types map[string]*quota.Type) error {
+	b.types = types
+	return nil
+}
+
+////////////////// Runtime Methods //////////////////////////
+
+func (h *handler) HandleQuota(context context.Context, instance *quota.Instance, args adapter.QuotaRequestArgs) (adapter.QuotaResult2, error) {
+	q := h.limits[instance.Name]
```
Comment (ZackButcher) [inline adapter/memquota2/memquota.go]: What if the instance name isn't found at runtime?

File: adapter/memquota2/config/config.proto (hunk 1 above)
Comment (geeknoid) [reply adapter/memquota2/config/config.proto]: Renamed to QuotaInfo

File: adapter/memquota2/memquota.go (hunk 3 above)
Comment (geeknoid) [reply adapter/memquota2/memquota.go]: There's a check in the Build method to ensure we have a limit for every instance we're receiving.

In general, I don't think we need adapters to double-check that Mixer isn't doing something invalid such as sending instances the adapters aren't prepared to handle.

File: adapter/memquota2/memquota.go (hunk 2 above)
Comment (geeknoid) [reply adapter/memquota2/memquota.go]: This follows the layout of all the other adapters I created. I was going to suggest changing your adapters to move GetBuilderInfo to the end of the file :-)

I find this bit of boilerplate not that interesting which is why I put it at the end, out of the way. I was actually thinking about moving this out into a standard info.go file to further isolate it...

Comment (ZackButcher) [reply adapter/memquota2/memquota.go]: I'll move my stuff down to the bottom I guess :)

**Pull request merged**